core/src/Kokkos_Macros.hpp

```diff
@@ -563,6 +567,29 @@ static constexpr bool kokkos_omp_on_host() { return false; }
 #endif
 #endif
 
+#ifdef KOKKOS_ENABLE_OPENACC
+#ifdef KOKKOS_COMPILER_NVHPC
+#define KOKKOS_IF_ON_DEVICE(CODE)   \
+  if (__builtin_is_device_code()) { \
+    KOKKOS_IMPL_STRIP_PARENS(CODE)  \
+  }
+#define KOKKOS_IF_ON_HOST(CODE)      \
+  if (!__builtin_is_device_code()) { \
+    KOKKOS_IMPL_STRIP_PARENS(CODE)   \
+  }
+#else
+// FIXE acc_on_device is a non-constexpr function
+#define KOKKOS_IF_ON_DEVICE(CODE)                     \
+  if constexpr (acc_on_device(acc_device_not_host)) { \
+    KOKKOS_IMPL_STRIP_PARENS(CODE)                    \
+  }
+#define KOKKOS_IF_ON_HOST(CODE)                   \
+  if constexpr (acc_on_device(acc_device_host)) { \
+    KOKKOS_IMPL_STRIP_PARENS(CODE)                \
+  }
+#endif
```

Comment:
This does not actually compile

Reply:
This code is for other OpenACC compilers. In OpenACC, the only portable way to check whether a code is running on a host or a device is to use `acc_on_device` function, which can be converted into a compile-time constant by the OpenACC compiler. Unfortunately, this compile-time conversion is not guaranteed even if the input argument is a compile-time constant, and thus this code may not work in some OpenACC compilers (just best-effort approach for now).

Reply:
Do we know of any compiler that supports it?

Reply:
I have not checked it; the OpenACC spec says that it can be evaluated to a constant at compile time, but it is optional.